### PR TITLE
Add CompressionLevel Calculation for PQ

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
Currently, for product quantization, we set the calculated compression level to NOT_CONFIGURED. The main issue with this is that if a user sets up a disk-based index with PQ, no re-scoring will happen by default.

This change adds the calculation so that the proper re-scoring will happen. The formula is fairly straightforward =>
actual compression = (d * 32) / (m * code_size). Then, we round to the neareste compression level (because we only support discrete compression levels).

One small issue with this is that if PQ is configured to have compression > 32x, the value will be 32x. Functionally, the only issue will be that we may not be as aggressive on oversampling for on disk mode.

### Check List
- [X] New functionality includes testing.
- [X] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by navneet1v** [comment]
> One small issue with this is that if PQ is configured to have compression > 32x, the value will be 32x. Functionally, the only issue will be that we may not be as aggressive on oversampling for on disk mode.

should we allow more compression level?

File: src/main/java/org/opensearch/knn/index/mapper/CompressionLevel.java [hunk 1]
```diff
@@ -29,6 +29,8 @@ public enum CompressionLevel {
     x16(16, "16x", new RescoreContext(3.0f, false), Set.of(Mode.ON_DISK)),
     x32(32, "32x", new RescoreContext(3.0f, false), Set.of(Mode.ON_DISK));
 
+    public static final CompressionLevel MAX_COMPRESSION_LEVEL = CompressionLevel.x32;
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/mapper/CompressionLevel.java:32]: should we have 64 as a max compression level? I don't have a solid point to have 64 but I think have 1 more extra compression is always good.

**Review by navneet1v — COMMENTED** [review]
overall looks good to me. Just 1 minor comment.

**Review by heemin32 — COMMENTED** [review]
LGTM. Do we want to add integ test for it? Recall should be higher with rescoring than baseline.

**Review by heemin32 — APPROVED** [review]

File: src/main/java/org/opensearch/knn/index/mapper/CompressionLevel.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/mapper/CompressionLevel.java:32]: I think that makes sense. I guess for 64x, default for all dimensions should probably be 5x.

**Review by navneet1v — APPROVED** [review]

**Review by heemin32 — APPROVED** [review]

**Comment by jmazanec15** [comment]
@navneet1v we did here: https://github.com/opensearch-project/k-NN/issues/1779#user-content-appendix-b-baseline-rescore-experiments

**Comment by navneet1v** [comment]
> @navneet1v we did here: [#1779 (comment)](https://github.com/opensearch-project/k-NN/issues/1779#user-content-appendix-b-baseline-rescore-experiments)

Thanks. I completely forgot about that.

**Pull request merged**